#' Function to simulate BAMs from a fasta sequence input
#'
#' The function takes a fasta file input (i.e. has 1 to N fasta sequences in the the fasta file depending on copy number status), 
#' simulates Illumina reads using ART and aligns the reads using BWA mem
#' @param clone_fasta_filename The name of the fasta file name for BAM generation
#' @param art_bin Full path to the ART bin tool for Illumina read simulation
#' @param haploid_cov Sequencing coverage depth to be simulated per chromosome copy (numeric)
#' @param read_length Length of the Illumina reads to be simulated (integer, usually 150)
#' @param fragment_size Mean size of the sequencing library fragments to be simulated (integer)
#' @param fragment_size_sd Standard deviation of the size of the sequencing library fragments to be simulated (integer)
#' @param tmp_dir Name of the temporary directory for samtools run (character string, e.g. "TMP")
#' @param fq1 Name of the fastq (mate 1) filename to be generated by ART (character string) 
#' @param fq2 Name of the fastq (mate 2) filename to be generated by ART (character string)
#' @param bwa Full path to the BWA tool for read alignment
#' @param refseq Full path to the fasta reference file to be used by BWA 
#' @param bam Name of the BAM file to be generated (string)
#' @param ncores Number of cores to be used in running BWA and samtools
#' @param samtools Full path to the samtools bin
#' @param logfile Name of the logfile for the BWA run (string)
#' @param skip_art set to TRUE if ART run is already complete - files are expected in the working directory (default = FALSE)
#' @param skip_bwa set to TRUE if BWA run is already complete - files are expected in the working directory (default = FALSE)
#' @author naser.ansari-pour
#' @export

fasta2bam_simulate <- function(clone_fasta_filename,art_bin,haploid_cov,read_length,fragment_size,fragment_size_sd,tmp_dir,fq1,fq2,bwa,refseq,bam,ncores,samtools,logfile,skip_art=FALSE,skip_bwa=FALSE){
  clone_fastq_filename=gsub(".fa","_",clone_fasta_filename)
  if (!skip_art){
    system(paste0(art_bin," -i ",clone_fasta_filename," -p -ss HS25 -f ",haploid_cov," -na -l ",read_length," -m ",fragment_size," -s ",fragment_size_sd," -o ",clone_fastq_filename), wait = TRUE)
  }
  if (!skip_bwa){
    system(paste("mkdir -p",tmp_dir))
    print(paste(fq1,fq2,bam))
    system(paste(bwa,"mem",refseq,fq1,fq2,"-t",ncores,"|",samtools,"sort - -O bam -o",bam,"-T",tmp_dir,"-@",ncores,">",logfile,"2>&1"), wait=TRUE)
    system(paste(samtools,"index",bam), wait = TRUE)
  }
}
